Simple ConnectionStateDisplay prefab

RaduDumitriu-Dev opened the pull request:
- adds ConnectionDisplay prefab
- implements it into existing ExampleScene
- requires the playthrough to be assigned to work
- currently only supports one, will likely need to be expanded to support multiple concurrent ones
- prefab is set up as a RectTransform(canvas bound) - may need a more debug oriented implementation (via OnGUI maybe?)

Closes #11

RaduDumitriu-Dev committed:
Simple ConnectionStateDisplay prefab

- adds ConnectionDisplay prefab
- implements it into existing ExampleScene
- requires the playthrough to be assigned to work
- currently only supports one, will likely need to be expanded to support multiple concurrent ones
- prefab is set up as a RectTransform(canvas bound) - may need a more debug oriented implementation (via OnGUI maybe?)

bensalilijames reviewed (changes requested):
I think we should approach this slightly differently - it would be great for connection state to be implemented within the `Playthrough` class as a native concept, some examples:

- Unreal: `ECharismaPlaythroughConnectionState` in https://github.com/charisma-ai/charisma-sdk-unreal/blob/main/Source/CharismaModule/Private/Playthrough.cpp
- JS: `ConnectionStatus` in https://github.com/charisma-ai/charisma-sdk-js/blob/main/src/Playthrough.ts

So we'd define a connection status enum there, then we need to track the following state changes:

- `Disconnected`: the initial state
- `Connecting`: when the client calls `Connect`
- `Connected`: after the `status: ready` event is received in `OnMessage("status")`
- `Disconnected`: upon `OnLeave` firing

We can add these as event callbacks that you can subscribe to in the `Playthrough`, so a consumer can do:

```c#
playthrough.OnConnectionStateChange += (newConnectionState) => {
  // do something
}
```

then we can remove the `Update` handler in `ConnectionStateDisplay` (quite wasteful to do this every frame) and subscribe to the callback instead!

happy to chat through if that's confusing 🙂

RaduDumitriu-Dev committed:
update to Connection state

- moved over connection state into Playthrough class
- added OnConnectionStateChange delegate for listening
- changed IsReadyToPlay Property to private function, in case functionality needs to change

Notes:
- IsReadyToPlay currently just queries the ConnectionState, but in the Unreal implementation it also checks if the current _room is valid/exists. can potentially modify this in the future. at time of writing it works the same as the old implementation
- Property IsConnected hasn't been removed as it ties into the _room. I think it does overlap in functionality with the ConnectionState, however I am not sure what the implications are of replacing it. Will need to have a deeper dive when looking at Reconnection behaviour most likely

RaduDumitriu-Dev committed:
updating Example Script sample

missed this file in previous commit

RaduDumitriu-Dev committed:
added review feedback

- updated ConnectionState naming to conventions
- added prints for Connecting/Disconnecting from a already connected/disconnected playthrough, to inform the user
- removed IsConnected property, changed it to a private function. its functionality is the same as IsReadyToPlay, however this may change in the future as the meaning is not the same and implementations may change.

bensalilijames commented on Scripts/Playthrough.cs:
```diff
@@ -182,14 +197,18 @@ public async void Connect(Action onReadyCallback)
         // Disconnect from the current interaction.
```
Sometimes the server will disconnect without the client asking it to (i.e. without the client calling `Disconnect`), in order to listen for this we should subscribe to `room.OnLeave()` (just below the `room.OnMessage` listeners above), and if this fires, it should `SetConnectionState(ConnectionState.Disconnected)` too

bensalilijames commented on Scripts/Playthrough.cs:
```diff
@@ -331,5 +350,29 @@ public void Play()
+
+        #region Private functions
+
+        private bool IsReadyToPlay()
+        {
+            return _connectionState == ConnectionState.Connected;
+        }
```
Duplicates `IsConnected` below, suggest removing `IsReadyToPlay` in favour of `IsConnected`

RaduDumitriu-Dev committed:
addressing feedback

- removed IsReadyToPlay in favor of IsConnected
- added room.OnLeave callback, to mark the playthrough as disconnected

RaduDumitriu-Dev committed:
Updated example scene to fix unassigned connectionState field

bensalilijames reviewed (approved):
looks great, thanks! 💯
